apollo-federation/src/query_graph/extract_subgraphs_from_supergraph.rs

```diff
@@ -2094,10 +2104,181 @@ fn maybe_dump_subgraph_schema(subgraph: FederationSubgraph, message: &mut String
     };
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// @join__directive extraction
+
+static JOIN_DIRECTIVE: &str = "join__directive";
+
+/// Converts `@join__directive(graphs: [A], name: "foo")` to `@foo` in the A subgraph.
+/// If the directive is a link directive on the schema definition, we also need
+/// to update the metadata and add the imported definitions.
+fn extract_join_directives(
+    supergraph_schema: &FederationSchema,
+    subgraphs: &mut FederationSubgraphs,
+    graph_enum_value_name_to_subgraph_name: &IndexMap<Name, Arc<str>>,
+) -> Result<(), FederationError> {
+    let join_directives = match supergraph_schema
+        .referencers()
+        .get_directive(JOIN_DIRECTIVE)
+    {
+        Ok(directives) => directives,
+        Err(_) => {
+            // No join directives found, nothing to do.
+            return Ok(());
+        }
+    };
+
+    if let Some(schema_def_pos) = &join_directives.schema {
+        let schema_def = schema_def_pos.get(supergraph_schema.schema());
+        let directives = schema_def
+            .directives
+            .iter()
+            .filter_map(|d| {
+                if d.name == JOIN_DIRECTIVE {
+                    Some(join_directive_to_real_directive(d))
+                } else {
+                    None
+                }
+            })
+            .collect_vec();
+
+        // TODO: Do we need to handle the link directive being renamed?
+        let (links, others) = directives
+            .into_iter()
+            .partition::<Vec<_>, _>(|(d, _)| d.name == DEFAULT_LINK_NAME);
+
+        // After adding links, we'll check the link against a safelist of
+        // specs and check_or_add the spec definitions if necessary.
+        for (link_directive, subgraph_enum_values) in links {
+            for subgraph_enum_value in subgraph_enum_values {
+                let subgraph = get_subgraph(
+                    subgraphs,
+                    graph_enum_value_name_to_subgraph_name,
+                    &subgraph_enum_value,
+                )?;
+
+                schema_def_pos.insert_directive(
+                    &mut subgraph.schema,
+                    Component::new(link_directive.clone()),
+                )?;
+
+                // TODO: add imported definitions from relevant specs
```

Comment:
can you add the todos to jira as well? we lose track of these so easily

Reply:
https://apollographql.atlassian.net/browse/FED-373